File: hydra-scala/src/main/scala/hydra/lib/lists/Lists.scala
package hydra.lib.lists


def concat[A](lists: Seq[Seq[A]]): Seq[A] = lists.fold(Seq.empty)((x, y) => x ++ y)

def head[A](l: Seq[A]): A = l.head

def intersperse[A](sep: A, l: Seq[A]): Seq[A] = (sep, l) match {
  case (_, Nil) => Nil
  case (_, Seq(x)) => Seq(x)
  case (s, y::ys) => y+:s+:intersperse(s, ys)
}

def intercalate[A](sep: Seq[A], l: Seq[Seq[A]]): Seq[A] = concat(intersperse(sep, l))

def last[A](l: Seq[A]): A = l.last

def length[A](l: Seq[A]): Int = l.length

def map[A, B](f: A => B, l: Seq[A]): Seq[B] = l.map(f)

File: hydra-haskell/src/main/haskell/Hydra/Impl/Haskell/Dsl/Lib/Lists.hs
module Hydra.Impl.Haskell.Dsl.Lib.Lists where

import Hydra.Phantoms
import qualified Hydra.Impl.Haskell.Dsl.Terms as Terms
import Hydra.Impl.Haskell.Sources.Libraries


concat :: Datum ([a] -> a)
concat = Datum $ Terms.primitive _lists_concat

head :: Datum ([a] -> a)
head = Datum $ Terms.primitive _lists_head

intercalate :: Datum ([a] -> [a] -> [a])
intercalate = Datum $ Terms.primitive _lists_intercalate

intersperse :: Datum ([a] -> a -> [a])
intersperse = Datum $ Terms.primitive _lists_intersperse

last :: Datum ([a] -> a)
last = Datum $ Terms.primitive _lists_last

length :: Datum ([a] -> Int)
length = Datum $ Terms.primitive _lists_length

--map :: Datum ((a -> b) -> [a] -> [b])
--map = Datum $ Terms.primitive _lists_map
